### PR TITLE
SmartContract: add FindOptions.Backwards to iterate in reverse order

### LINKED ISSUE #2789: Add a FindOption to control SeekDirection of Storage.Find
**Summary or problem description**
Storage.Find contract API exposed by the system allows to iterate over a set of keys with various options, but it only does so in one direction (ascending) even though the DataCache/Snapshot behind the scenes has a definition of SeekDirection and therefore is able to seek through the elements in descending order. Depending on the contract's storage schema and the nature of the data stored there might be a need to get X number of the latest values from a set, currently this requires iterating over the whole set till it ends.

**Do you have any solution you want to propose?**
Add a new flag to FindOptions to control the SeekDirection and allow to iterate over a set of KV pairs in descending (backwards) order.

**Neo Version**
- Neo 3

**Where in the software does this update applies to?**
- SDK

### PR DESCRIPTION
**Pull request by roman-khimov**
Depending on the data scheme contracts may want to iterate over prefix in descending order. Fixes #2789.

### CODE REVIEW

File: src/Neo/SmartContract/ApplicationEngine.Storage.cs [hunk 1]
```diff
@@ -152,7 +153,10 @@ protected internal IIterator Find(StorageContext context, byte[] prefix, FindOpt
             if ((options.HasFlag(FindOptions.PickField0) || options.HasFlag(FindOptions.PickField1)) && !options.HasFlag(FindOptions.DeserializeValues))
                 throw new ArgumentException(null, nameof(options));
             byte[] prefix_key = StorageKey.CreateSearchPrefix(context.Id, prefix);
-            return new StorageIterator(Snapshot.Find(prefix_key).GetEnumerator(), prefix.Length, options);
+            SeekDirection direction = SeekDirection.Forward;
+            if (options.HasFlag(FindOptions.Backwards))
+                direction = SeekDirection.Backward;
```
Comment (AnnaShaleva) [inline src/Neo/SmartContract/ApplicationEngine.Storage.cs:158]: A little syntactic sugar:
```suggestion
            SeekDirection direction = options.HasFlag(FindOptions.Backwards) ? SeekDirection.Backward : SeekDirection.Forward;
```

**Review by AnnaShaleva — DISMISSED** [review]

File: src/Neo/SmartContract/ApplicationEngine.Storage.cs (hunk 1 above)
Comment (roman-khimov) [reply src/Neo/SmartContract/ApplicationEngine.Storage.cs:158]: Sure. 3+ years of Go programming and you forget ternary operator exists.

**Review by AnnaShaleva — APPROVED** [review]

File: src/Neo/SmartContract/FindOptions.cs [hunk 2]
```diff
@@ -53,9 +53,14 @@ public enum FindOptions : byte
         /// </summary>
         PickField1 = 1 << 5,
 
+        /// <summary>
+        /// Indicates that results should be returned in backwards (descending) order.
+        /// </summary>
+        Backwards = 1 << 6,
```
Comment (erikzhang) [inline src/Neo/SmartContract/FindOptions.cs:59]:
```suggestion
        Backwards = 1 << 7,
```

**Comment by superboyiii** [comment]
@erikzhang @roman-khimov  Let's move on.

**Comment by Liaojinghui** [comment]
@shargon Mind to review this on?

File: src/Neo/Persistence/DataCache.cs [hunk 3]
```diff
@@ -202,13 +202,37 @@ public void Delete(StorageKey key)
...
+                for (int i = key_prefix.Length - 1; i >= 0; i--)
+                {
+                    if (key_prefix[i] < 0xff)
+                    {
+                        seek_prefix = key_prefix.Take(i + 1).ToArray();
+                        // The next key after the key_prefix.
+                        seek_prefix[i]++;
+                        break;
+                    }
+                }
```
Comment (shargon) [inline src/Neo/Persistence/DataCache.cs:226]: Why the seek prefix it's different if is `Backward`?

Comment (roman-khimov) [reply src/Neo/Persistence/DataCache.cs:226]: See the example in UT, suppose we have "jjj", "key1", "key2" and "key3" items in the DB and we want to search for items with the "key" prefix. So to search in forward direction we place the cursor at the "key" item (nonexistent, but anyway) and move forward, retreiving items we want:
```
jjj
     <- cursor at "key"
key1
key2
key3
```
If you're to just move in the opposite direction you'll get a `jjj` item (or nothing if we're to check for the prefix additionally. So we do something like this here:
```
jjj
key1
key2
key3
     <- cursor at "keyz"
```
And get "key3", "key2" and "key1".

**Review by shargon — DISMISSED** [review]

**Review by AnnaShaleva — DISMISSED** [review]

**Comment by AnnaShaleva** [comment]
@Liaojinghui, @superboyiii, @erikzhang, could you review this PR, please? Let's move on with it, it's pretty useful.

File: tests/Neo.UnitTests/IO/Caching/UT_DataCache.cs [hunk 4]
```diff
@@ -152,11 +152,40 @@ public void TestFind()
...
+            items = myDataCache.Find(prefix, SeekDirection.Backward);
+            items.Count().Should().Be(4);
+            key4.Should().Be(items.ElementAt(0).Key);
+            value4.EqualsTo(items.ElementAt(0).Value).Should().BeTrue();
+            key3.Should().Be(items.ElementAt(1).Key);
+            value3.EqualsTo(items.ElementAt(1).Value).Should().BeTrue();
+            key2.Should().Be(items.ElementAt(2).Key);
+            value2.Should().Be(items.ElementAt(2).Value);
+            key1.Should().Be(items.ElementAt(3).Key);
+            value1.Should().Be(items.ElementAt(3).Value);
```
Comment (vncoelho) [inline tests/Neo.UnitTests/IO/Caching/UT_DataCache.cs:187]: tests for the null cases length 0?

**Review by Liaojinghui — APPROVED** [review]

File: tests/Neo.UnitTests/IO/Caching/UT_DataCache.cs (hunk 4 above)
Comment (roman-khimov) [reply tests/Neo.UnitTests/IO/Caching/UT_DataCache.cs:199]: Fixed and improved impl. a bit.

**Review by Liaojinghui — APPROVED** [review]

**Review by vncoelho — APPROVED** [review]

**Pull request merged**